File: FluidStudio/PGCore/Math/Primitive3d/Cone3d.cs
﻿using System;

namespace PG.Core.Math
{
    public class Cone3d : ISurface3d, IVolume3d
    {
        public Cone3d()
        {
            Radius = 1.0;
            Height = 1.0;
            Bottom = new Vector3d(0, 0, 0);
        }

        public Cone3d(double radius, double height, Vector3d bottom)
        {
            Radius = radius;
            Height = height;
            Bottom = bottom;
        }

        public double Radius { get; }

        public double Height { get; }

        public Vector3d Bottom { get; }

        public Vector3d GetPosition(double u, double v)
        {
            return GetPosition(u, v, 1.0);
        }

        public Vector3d GetPosition(double u, double v, double w)
        {
            var circle = ToCircleAt(w);
            return circle.GetPosition(u, v);
        }

        public bool IsInside(Vector3d position)
        {
            throw new NotImplementedException();
        }

        public Circle3d ToCircleAt(double p)
        {
            var radius = Radius * (1.0 - p);
            var z = p * Height;
            var center = Bottom + new Vector3d(0,0,z);
            return new Circle3d(radius, center);
        }
    }
}

File: CGLib/Crystal/Math/Cone3d.cpp
#include "Cone3d.h"

#include "Tolerance.h"

using namespace Crystal::Math;

Cone3d::Cone3d() :
	bottom(0, 0, 0),
	radius(1.0),
	height(1.0)
{}

Cone3d::Cone3d(const Vector3dd& bottom, const double radius, const double height) :
	bottom(bottom),
	radius(radius),
	height(height)
{
}

Vector3dd Cone3d::getPosition(const double u, const double v) const
{
	const auto uu = u * 2.0 * PI;
	const auto r = radius * (1.0 - v);
	const auto x = r * ::cos(uu);
	const auto y = r * ::sin(uu);
	const auto z = v * height;
	return bottom + Vector3dd(x, y, z);
}

Vector3dd Cone3d::getPosition(const double u, const double v, const double w) const
{
	const auto uu = u * 2.0 * PI;
	const auto r = w * radius * (1.0 - v);
	const auto x = r * ::cos(uu);
	const auto y = r * ::sin(uu);
	const auto z = v * height;
	return bottom + Vector3dd(x, y, z);
}

Vector3dd Cone3d::getNormal(const double u, const double v) const
{
	const auto dx = 1.0e-3;
	const auto v0 = getPosition(u, v);
	const auto v1 = getPosition(u + dx, v);
	const auto v2 = getPosition(u, v + dx);
	return glm::cross(v1 - v0, v2 - v0);
}

bool Cone3d::isInside(const Vector3dd& position) const
{
	assert(false);
	return false;
}
